Week 7: Lesson 7 Assignment - Admin Dashboard & Course Display

ltphongssvn opened the pull request:
## Summary
Complete implementation of Lesson 7 Assignment with both required features:
- **Question 1**: Admin dashboard with upcoming trimester display
- **Question 2**: Course display with enrolled students validation

## Assignment Implementation

### Question 1: Admin Dashboard with Upcoming Trimester Display
- ✅ Created `AdminDashboardController` with dynamic trimester logic
- ✅ Added GET `/dashboard` route mapping to `admin_dashboard#index`
- ✅ Implemented upcoming trimester query (start_date < 6 months from today)
- ✅ Updated admin dashboard view with ERB for dynamic term-year display
- ✅ Added comprehensive request specs with proper test data setup

### Question 2: Course Display with Enrolled Students
- ✅ Verified existing resourceful route `GET /courses/:id` (correct approach)
- ✅ Confirmed course show functionality displays course name and student names
- ✅ Added request specs testing course name and enrolled student display
- ✅ Validated existing implementation meets assignment requirements

## Technical Changes
- **New Controller**: `app/controllers/admin_dashboard_controller.rb`
- **Route Addition**: Dashboard route added to existing resourceful structure
- **View Enhancement**: Dynamic ERB templating for trimester display
- **Test Coverage**: Complete request specs for both assignment questions

## Test Results
- **Total Examples**: 2 examples, 0 failures
- **Admin Dashboard Spec**: Tests upcoming trimester display functionality
- **Course Spec**: Tests course name and enrolled student display
- **Coverage**: Complete assignment requirements validated

**Status**: Ready for code review

inlibra-cpa commented on app/controllers/courses_controller.rb:
```diff
@@ -47,13 +52,16 @@ def destroy
-end
+  # Only allow a list of trusted parameters through.
+  def course_params
+    params.expect(course: [ :coding_class_id, :trimester_id, :max_enrollment ])
```
Using `params.expect` (not a Rails API) will raise and bypass strong params(same for `set_course`).

Here's the standard solution example:
```
def course_params
  params.require(:course).permit(:coding_class_id, :trimester_id, :max_enrollment)
end
```

inlibra-cpa commented on app/controllers/trimesters_controller.rb:
```diff
@@ -0,0 +1,97 @@
+  # PATCH/PUT /trimesters/1 or /trimesters/1.json
+  def update
+    # Check if application_deadline parameter is missing
+    if params[:trimester][:application_deadline].blank?
```
`params[:trimester][:application_deadline]` will `NoMethodError` if `:trimester` is missing. You might want to use `params.dig(:trimester, :application_deadline)` in checks.

inlibra-cpa commented on app/models/trimester.rb:
```diff
@@ -1,7 +1,14 @@
   validates :application_deadline, presence: true
+
+  def display_name
+    "#{term}-#{year}"
```
The assignment asked for space-separated ("Fall 2025"). Update unless your tests explicitly require the hyphen.

inlibra-cpa reviewed (changes requested):
Hi @ltphongssvn ,
Week 9 asks for (1) an edit-page spec that asserts the "application deadline" label, and (2) PUT specs for happy path, missing deadline → 400, invalid date → 400, and missing ID → 404. Please add them.
Please update `params.expect` → `require/permit`, remove the duplicate courses routes, harden the trimester parameter checks, align `display_name` with the spec, and add the Week-9 request specs. Then this PR can be merged.

ltphongssvn commented:
@inlibra-cpa Thank you for your thorough code review. I've addressed all your feedback in commit b5ca7e1. Here's what I've updated:

## Controller Changes
- **courses_controller.rb**: Replaced `params.expect` with the standard `params.require(:course).permit(...)` pattern in the `course_params` method
- **courses_controller.rb**: Updated `set_course` method to use conventional `params[:id]` instead of `params.expect(:id)`
- **trimesters_controller.rb**: Fixed the parameter access to use `params.dig(:trimester, :application_deadline)` for safe nested hash navigation, preventing NoMethodError when :trimester key is missing

## Model Changes
- **trimester.rb**: Corrected the `display_name` method to use space separator as specified - now returns "Fall 2025" instead of "Fall-2025"

## Test Coverage
Added comprehensive request specs in `spec/requests/trimesters_spec.rb` with all the tests you requested:
- Edit page spec that asserts the "application deadline" label is present
- PUT request happy path test for successful updates
- PUT request test for missing deadline returning 400 Bad Request
- PUT request test for invalid date format returning 400 Bad Request
- PUT request test for non-existent ID returning 404 Not Found

All 5 RSpec tests are passing successfully (confirmed with `bundle exec rspec spec/requests/trimesters_spec.rb`).

The routes configuration was already clean with no duplicates - `resources :courses` and `resources :trimesters` are each defined only once.

These changes improve the code's robustness, follow Rails conventions more closely, and ensure proper error handling for edge cases. Please let me know if you need any additional changes or have questions about the implementation.

shaun-technovation commented on app/views/courses/_form.html.erb:
```diff
@@ -13,12 +14,12 @@
   <div>
     <%= form.label :trimester_id, style: "display: block" %>
-    <%= form.text_field :trimester_id %>
+    <%= form.collection_select :trimester_id, Trimester.all, :id, :display_name, { prompt: "Select a trimester" } %>
```
👍

shaun-technovation commented on app/views/courses/_form.html.erb:
```diff
@@ -13,12 +14,12 @@
   <div>
     <%= form.label :coding_class_id, style: "display: block" %>
-    <%= form.text_field :coding_class_id %>
+    <%= form.collection_select :coding_class_id, CodingClass.all, :id, :title, { prompt: "Select a coding class" } %>
```
👍

shaun-technovation commented on app/views/trimesters/edit.html.erb:
```diff
@@ -0,0 +1,50 @@
+<%# File: app/views/trimesters/edit.html.erb %>
+<h1>Edit Trimester</h1>
```
👍

shaun-technovation commented on spec/requests/trimesters_spec.rb:
```diff
@@ -0,0 +1,48 @@
+require 'rails_helper'
+
+RSpec.describe "Trimesters", type: :request do
```
Nice specs!

shaun-technovation reviewed (approved):
Awesome, great work here! Also, nice overview/summaries in all of your PRs! 💯 👏

ltphongssvn commented:
Thank you, I appreciate it!
